src/Streamly/Internal/Data/Fold/Types.hs

```diff
@@ -975,6 +982,69 @@ many (Fold sstep sinitial sextract) (Fold cstep cinitial cextract) =
     -- Caution! There is mutual recursion here, inlining the right functions is
     -- important.
 
+    {-# INLINE handleSplitStep #-}
+    handleSplitStep branch cs sres =
+        case sres of
+            Partial ss1 -> return $ Partial $ branch ss1 cs
+            Done sb -> runCollector ManyFirst cs sb
+
+    {-# INLINE handleCollectStep #-}
+    handleCollectStep branch cres =
+        case cres of
+            Partial cs -> do
+                sres <- sinitial
+                handleSplitStep branch cs sres
+            Done cb -> return $ Done cb
+
+    -- Do not inline this
+    runCollector branch cs sb = cstep cs sb >>= handleCollectStep branch
+
+    initial = cinitial >>= handleCollectStep ManyFirst
+
+    {-# INLINE step_ #-}
+    step_ ss cs a = do
+        sres <- sstep ss a
+        handleSplitStep ManyLoop cs sres
+
+    {-# INLINE step #-}
+    step (ManyFirst ss cs) a = step_ ss cs a
+    step (ManyLoop ss cs) a = step_ ss cs a
+
+    extract (ManyFirst _ cs) = cextract cs
+    extract (ManyLoop ss cs) = do
+        sb <- sextract ss
+        cres <- cstep cs sb
+        case cres of
+            Partial s -> cextract s
+            Done b -> return b
+
+-- | Collect zero or more applications of a fold.  @many split collect@ applies
+-- the @split@ fold repeatedly on the input stream and accumulates zero or more
+-- fold results using @collect@.
```

Comment:
"Like many, but inner fold emits an output at the end even if no input is received."